File: dotnet/src/Mapping/ExtractStripeEntityRefs.cs
namespace Subscrio.AuditLog.Mapping;

/// <summary>
/// Stripe customer / subscription IDs pulled from a verified Stripe event payload.
/// </summary>
public sealed class StripeEntityRefs
{
    public string? StripeCustomerId { get; init; }
    public string? StripeSubscriptionId { get; init; }
}

/// <summary>
/// Best-effort extraction of Stripe entity IDs from Event.Data.Object.
/// </summary>
public static class ExtractStripeEntityRefs
{
    public static StripeEntityRefs FromEvent(Stripe.Event? stripeEvent)
    {
        if (stripeEvent?.Data?.Object == null)
            return new StripeEntityRefs();

        var obj = stripeEvent.Data.Object;
        var type = stripeEvent.Type ?? string.Empty;

        // customer.* (not customer.subscription.*)
        if (type.StartsWith("customer.", StringComparison.Ordinal)
            && !type.StartsWith("customer.subscription.", StringComparison.Ordinal))
        {
            return new StripeEntityRefs
            {
                StripeCustomerId = obj is Stripe.Customer c ? AsId(c.Id) : null
            };
        }

        // customer.subscription.*
        if (type.StartsWith("customer.subscription.", StringComparison.Ordinal)
            && obj is Stripe.Subscription subFromType)
        {
            return FromSubscription(subFromType);
        }

        return obj switch
        {
            Stripe.Invoice invoice => new StripeEntityRefs
            {
                StripeCustomerId = AsId(invoice.CustomerId) ?? AsId(invoice.Customer?.Id),
                StripeSubscriptionId = AsId(invoice.Parent?.SubscriptionDetails?.SubscriptionId)
                    ?? AsId(invoice.Parent?.SubscriptionDetails?.Subscription?.Id)
            },
            Stripe.Checkout.Session session => new StripeEntityRefs
            {
                StripeCustomerId = AsId(session.CustomerId) ?? AsId(session.Customer?.Id),
                StripeSubscriptionId = AsId(session.SubscriptionId) ?? AsId(session.Subscription?.Id)
            },
            Stripe.Customer customer => new StripeEntityRefs { StripeCustomerId = AsId(customer.Id) },
            Stripe.Subscription subscription => FromSubscription(subscription),
            _ => new StripeEntityRefs()
        };
    }

    private static StripeEntityRefs FromSubscription(Stripe.Subscription subscription) =>
        new()
        {
            StripeSubscriptionId = AsId(subscription.Id),
            StripeCustomerId = AsId(subscription.CustomerId) ?? AsId(subscription.Customer?.Id)
        };

    private static string? AsId(string? value) =>
        string.IsNullOrEmpty(value) ? null : value;
}

File: typescript/src/extractStripeEntityRefs.ts
/**
 * Pull Stripe customer / subscription IDs from a verified Stripe.Event payload.
 * Best-effort: unknown shapes leave fields undefined.
 */
export interface StripeEntityRefs {
  stripeCustomerId?: string;
  stripeSubscriptionId?: string;
}

function asId(value: unknown): string | undefined {
  if (typeof value === 'string' && value.length > 0) return value;
  if (
    value !== null &&
    typeof value === 'object' &&
    'id' in value &&
    typeof (value as { id: unknown }).id === 'string'
  ) {
    const id = (value as { id: string }).id;
    return id.length > 0 ? id : undefined;
  }
  return undefined;
}

export function extractStripeEntityRefs(stripeEvent: {
  type?: string | null;
  data?: { object?: unknown } | null;
} | null | undefined): StripeEntityRefs {
  if (!stripeEvent?.data?.object || typeof stripeEvent.data.object !== 'object') {
    return {};
  }

  const obj = stripeEvent.data.object as Record<string, unknown>;
  const type = stripeEvent.type ?? '';

  // customer.* (not customer.subscription.*)
  if (type.startsWith('customer.') && !type.startsWith('customer.subscription.')) {
    return { stripeCustomerId: asId(obj.id) };
  }

  // customer.subscription.*
  if (type.startsWith('customer.subscription.')) {
    return {
      stripeSubscriptionId: asId(obj.id),
      stripeCustomerId: asId(obj.customer),
    };
  }

  // invoice.*, checkout.session.*, charge.*, etc.
  const stripeCustomerId = asId(obj.customer);
  const parent = obj.parent as Record<string, unknown> | undefined;
  const subscriptionDetails = parent?.subscription_details as Record<string, unknown> | undefined;
  const stripeSubscriptionId = asId(obj.subscription) ??
    asId(subscriptionDetails?.subscription) ??
    (obj.object === 'subscription' ? asId(obj.id) : undefined);

  if (stripeCustomerId || stripeSubscriptionId) {
    return { stripeCustomerId, stripeSubscriptionId };
  }

  if (obj.object === 'customer') {
    return { stripeCustomerId: asId(obj.id) };
  }

  return {};
}
